### PR TITLE
add namespace function scanner

### PR DESCRIPTION
**Pull request by FredKSchott**
Functions tagged with `@memberof` are now returned by the new function scanner. Returning these functions via analysis is still TODO.

### CODE REVIEW

File: src/javascript/function-scanner.ts [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+ * http://polymer.github.io/CONTRIBUTORS.txt
+ * Code distributed by Google as part of the polymer project is also
+ * subject to an additional IP rights grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+
+import * as estree from 'estree';
+
+import {getIdentifierName, getNamespacedIdentifier} from
+'../javascript/ast-value';
```
Comment (justinfagnani) [inline src/javascript/function-scanner.ts]: `npm run format`?

File: src/javascript/function-scanner.ts [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+
+    const docs = jsdoc.parseJsdoc(comment);
+    const functionName = getNamespacedIdentifier(anaylzedName, docs);
+    const sourceRange = this.document.sourceRangeForNode(node)!;
+    const returnTag = jsdoc.getTag(docs, 'return');
+    const summaryTag = jsdoc.getTag(docs, 'summary');
+    const summary = (summaryTag && summaryTag.description) || '';
+    const description = docs.description;
+
+    let functionReturn;
```
Comment (justinfagnani) [inline src/javascript/function-scanner.ts]: `const` and ternary?

File: src/javascript/function-scanner.ts [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+      functionReturn = {
+        type: returnTag.type || undefined,
+        desc: returnTag.description || '',
+      };
+    }
+
+    let functionParams;
+    if (docs.tags) {
+      const paramTags = docs.tags.filter((tag) => tag.tag === 'param');
+      if (paramTags.length > 0) {
```
Comment (justinfagnani) [inline src/javascript/function-scanner.ts]: Why not just map anyway and always produce an array, even if empty?

File: src/javascript/function-scanner.ts (hunk 1 above)
Comment (FredKSchott) [inline src/javascript/function-scanner.ts]: hmm,  not sure how this got in there

File: src/javascript/function-scanner.ts (hunk 2 above)
Comment (FredKSchott) [inline src/javascript/function-scanner.ts]: Woah, clang-format does not like that:

```

    const functionReturn = (returnTag) ? {
      type: returnTag.type || undefined,
      desc: returnTag.description || '',
    } :
                                         undefined;
```

I'm going to stick with let, I agree const is great but not at the expense of significant clarity.

**Pull request merged**